### PR TITLE
Bugfix: Data shuffling

### PR DESCRIPTION
**Pull request by DriesSmit**
## What?
Fix a bug where we shuffle over both the batch and sequence dimensions in the trainer. This breaks the bootstrapped value estimation calculation and also limits the recurrent policy. 
## Why?
We should only shuffle over the batch dimension.
## How?
Update the trainer code.
## Extra
.

### CODE REVIEW

File: mava/components/training/model_updating.py [hunk 1]
```diff
@@ -295,24 +294,23 @@ def model_update_epoch(
...
-            )
+            permutation = jax.random.permutation(shuffle_key, batch_shape)
 
             shuffled_batch = jax.tree_util.tree_map(
                 lambda x: jnp.take(x, permutation, axis=0), batch
             )
+
+            # Flatten the batch.
+            shuffled_batch = jax.tree_util.tree_map(
+                lambda x: x.reshape((-1,) + x.shape[2:]), shuffled_batch
```
Comment (sash-a) [inline mava/components/training/model_updating.py:311]: Haiku have a great method for this: `merge_leading_dims`

It's in haiku._src I think

File: mava/components/training/model_updating.py [hunk 2]
```diff
@@ -295,24 +294,23 @@ def model_update_epoch(
...
-            # assert ...
-            assert (
-                list(batch.observations.values())[0].observation.shape[0]
-                == trainer.store.full_batch_size
-            )
+            # Note (dries): This computation is only performed once at compilation tome
+            # and not at every epoch step.
+            batch_shape = jax.tree_util.tree_leaves(
+                list(batch.observations.values())[0].observation
+            )[0].shape[0]
```
Comment (sash-a) [inline mava/components/training/model_updating.py:301]: Can leave it here, but might be cleaner to store the batch size in the store as we used to do with full_batch_size

**Review by sash-a — CHANGES_REQUESTED** [review]
Great find Dries, just minor code style things! :fire:

File: mava/components/training/model_updating.py (hunk 1 above)
Comment (DriesSmit) [reply mava/components/training/model_updating.py:311]: Awesome thanks :raised_hands: I was searching for that method.

File: mava/components/training/model_updating.py (hunk 2 above)
Comment (DriesSmit) [reply mava/components/training/model_updating.py:301]: Agreed. I removed it as it is the same size as sample_batch_size :+1:

File: mava/components/training/losses.py [hunk 3]
```diff
@@ -247,8 +248,9 @@ def policy_loss_fn(
                         )
 
                         # Flatten the distribution_params
+
                         distribution_params = jax.tree_util.tree_map(
-                            lambda x: x.reshape((-1,) + x.shape[2:]),
+                            lambda x: merge_leading_dims(x, 2),
```
Comment (RuanJohn) [inline mava/components/training/losses.py:253]: I really like this!

Comment (DriesSmit) [reply mava/components/training/losses.py:253]: I was @sash-a 's idea :raised_hands:

**Review by sash-a — APPROVED** [review]
Looks great! Hopefully the benchmarks are good! :fire:

**Review by RuanJohn — APPROVED** [review]
Thank you @DriesSmit! And a very good catch! 
Approved, just pending the benchmark results.
